### ISSUE TITLE
False positive of `reportUnnecessaryComparison` for loop variables

### ISSUE THREAD

**Issue by ghost** [microsoft/pyright#10311]
**Describe the bug & Code or Screenshots**
In the following code, the loop condition `i != 10` is flagged by Pyright, while `j != 10` is accepted, even though both loops implement the same logic ([PP](https://pyright-play.net/?strict=true&code=CYUwZgBALiDOUAoCUEC0A%2BCAjA9jgNgFwQB0ZAsAFBVUCWEAvBAAxUDuAFrfiBPQIRMAjM2IQAxBAAKATwBOtAOYcoxAMI4AdsFpRaWiG274IAQ3xtTM2BBAA3cwFdTMaDggAVOY96xamgGNeKA5gmQAHOAgAIgAZXRA5cwBtZgBdaLNtGPiYJPxkkQyIDlM7Xk13HHK5fFNwqggmvkgYeGRCRubugK09TR8upvoAamEaSgArRhZ2Lh4IacEIEU7KbtpWuEQkNe6evv9B9ebpsZWqIA)):

```python
def test() -> bool: ...

i = 0
while i != 10:  # Pyright: Condition will always evaluate to True since the types "Literal[0]" and "Literal[10]" have no overlap
    if test():
        continue
    i += 1

j = 0
while j != 10:
    if test():
        continue
    j += 1
```

**VS Code extension or command-line**
Pylance version: 2025.4.100

**Comment by erictraut** [comment]
Thanks for the bug report and the clean, simple repro case. I have some theories about the underlying cause, but I'll need to investigate further.

**Comment by erictraut** [comment]
The second example is working as intended, so I don't consider that a bug. Pyright performs type narrowing locally within an execution context without regard for side effects that may occur in other execution contexts. Static type checkers cannot generally determine if and when code within other contexts may be executed. If you want your code to work well with static analyzers, avoid writing code that has side effects like this.

**Comment by RomanValov** [comment]
I have bumped into similar issue with following code:

```python
class Test:

    def __init__(self) -> None:
        self._value: int | None = 0

    @property
    def value(self) -> int | None:
        return self._value

    def enact(self) -> None:
        self._value = None


if __name__ == '__main__':

    test = Test()
    if test.value is None:
        raise RuntimeError

    while test.value is not None:  # Condition will always evaluate to True since the types "int" and "None" have no overlap
        test.enact()
```

On one hand the snippet also involves side effects. But on the other `test.value` is a property -- basically a function call. It's result should not be considered for type narrowing.
